stores/env/store.go

```diff
@@ -0,0 +1,119 @@
+package env //import "go.mozilla.org/sops/stores/env"
+import (
+	"bufio"
+	"bytes"
+	"encoding/base64"
+	"encoding/gob"
+	"fmt"
+	"go.mozilla.org/sops"
+	"go.mozilla.org/sops/stores"
+	"strings"
+)
+
+// Store handles storage of env data
+type Store struct {
+}
+
+func (store *Store) LoadEncryptedFile(in []byte) (sops.Tree, error) {
+	branch, err := store.LoadPlainFile(in)
+	if err != nil { return sops.Tree{}, err }
+
+	storeMetadata := stores.Metadata{}
+	index := -1
+	for i, item := range branch {
+		if item.Key == "_metadata" {
+			storeMetadata, err = FromGOB64(fmt.Sprint(item.Value))
+			if err != nil { return sops.Tree{}, err }
+			index = i
+			break
+		}
+	}
+	if index == -1 { return sops.Tree{}, sops.MetadataNotFound }
+
+	internalMetadata, err := storeMetadata.ToInternal()
+	if err != nil { return sops.Tree{}, err }
+
+	return sops.Tree{
+		Branch:   append(branch[:index], branch[index+1:]...),
+		Metadata: internalMetadata,
+	}, nil
+}
+
+func (store *Store) LoadPlainFile(in []byte) (sops.TreeBranch, error) {
+	branch := make(sops.TreeBranch, 0)
+	reader := bytes.NewReader(in)
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "" { continue }
+		pos := strings.Index(line, "=")
+		if pos == -1 {
+			// FIXME: Print line number
+			return nil, fmt.Errorf("could not parse line: %s", line)
+		}
+		// FIXME: Trim key and value? Remove quotation marks?
+		branch = append(branch, sops.TreeItem{
+			Key:   line[:pos],
+			Value: line[pos+1:],
+		})
+	}
+	return branch, nil
+}
+
+func (store *Store) EmitEncryptedFile(in sops.Tree) ([]byte, error) {
+	plain, err := store.EmitPlainFile(in.Branch)
+	if err != nil { return nil, err }
+	buffer := bytes.NewBuffer(plain)
+	metadata := stores.MetadataFromInternal(in.Metadata)
+	str, err := ToGOB64(metadata)
+	if err != nil { return nil, err }
+	line := fmt.Sprintf("_metadata=%s\n", str)
+	buffer.WriteString(line)
+	return buffer.Bytes(), nil
+}
+
+func (store *Store) EmitPlainFile(in sops.TreeBranch) ([]byte, error) {
+	buffer := bytes.Buffer{}
+	for _, item := range in {
+		// FIXME: Check that item.Value is a scalar.
+		// FIXME: Does Go know how to print the OS-specific EOL string? Do we care?
+		line := fmt.Sprintf("%s=%s\n", item.Key, item.Value)
+		buffer.WriteString(line)
+	}
+	return buffer.Bytes(), nil
+}
+
+func (Store) EmitValue(v interface{}) ([]byte, error) {
+	// FIXME: Whot should this function do?
+	panic("implement me")
+}
+
```

Comment:
The code below is an attempt to encode the metadata. Is this the best way?

Reply:
It is a way :) It has the downside that it's not obvious what options the file has, which is easy to see with the JSON/YAML implementations currently. Maybe it'd be possible to instead "flatten" the metadata struct into fields, e.g.

```
SOPS_VERSION=3.1.1
SOPS_KMS_KEY_1=blah
```

But this can get very annoying to do.

Reply:
I'll see how far I can carry that idea.

Reply:
I used `_metadata`, but I think `sops` is better because 1) lowercase is often ignored by parsers of dotenv files, 2) it is also used by the YAML and JSON file types.

It is relatively easy to encode the string field of [Metadata](https://github.com/mozilla/sops/blob/7c72237f5a49f7b8815efababe55e00e60fecf4f/sops.go#L425) into `sops_version` etc. (See the current code.) However, writing this out for the structured data does not seem to be the way to go. Those raw strings at the end of the metadata are starting to make sense. :-) I think there is a stable translation possible like you suggest but I don't know how to write something like that in Go.

Reply:
Yeah, flattening the structure isn't exactly trivial. Since the main concern here is being able to easily see the metadata with a normal text editor, maybe it would be better to just encode the metadata to JSON (basically the same thing the JSON store does) and shove it into a field in the dotenv file. It's not great, because it will all be in one line, but imo it's better than having an essentially unreadable base64. Many differs can handle same-line diffs just fine, so I think it's at least an improvement.

Reply:
Your comment got me thinking. It's possible to [convert the metadata struct to a map](https://stackoverflow.com/a/42849112). I can write a function that converts that to flat key/value items and back. I'll give that a try and if I get stuck I'll just do a single JSON item.

Reply:
Yeah, but I think you'll run into the same issue. The map won't be a flat structure either.